src/adapter/src/coord/timeline.rs

```diff
@@ -82,93 +79,59 @@ impl<T: CoordTimestamp> TimestampOracle<T> {
         F: Fn() -> T + 'static,
     {
         Self {
-            state: TimestampOracleState::Writing(initially.clone()),
-            advance_to: Some(initially),
+            state: TimestampOracleState {
+                read_ts: initially.clone(),
+                write_ts: initially,
+            },
             next: Box::new(next),
         }
     }
 
-    /// Peek the current value of the timestamp.
-    ///
-    /// No operations should be assigned to the timestamp returned by this function. The
-    /// timestamp returned should only be used to compare the progress of the `TimestampOracle`
-    /// against some external source of time.
-    ///
-    /// Subsequent values of `self.read_ts()` and `self.write_ts()` will be greater or equal to
-    /// this timestamp.
-    ///
-    /// NOTE: This can be removed if DDL is included in group commits.
-    fn peek_ts(&self) -> T {
-        match &self.state {
-            TimestampOracleState::Writing(ts) | TimestampOracleState::Reading(ts) => ts.clone(),
-        }
-    }
-
     /// Acquire a new timestamp for writing.
     ///
     /// This timestamp will be strictly greater than all prior values of
-    /// `self.read_ts()`, and less than or equal to all subsequent values of
     /// `self.read_ts()`.
-    fn write_ts(&mut self) -> T {
-        match &self.state {
-            TimestampOracleState::Writing(ts) => ts.clone(),
-            TimestampOracleState::Reading(ts) => {
-                let mut next = (self.next)();
-                if next.less_equal(&ts) {
-                    next = ts.step_forward();
-                }
-                assert!(ts.less_than(&next));
-                self.state = TimestampOracleState::Writing(next.clone());
-                self.advance_to = Some(next.clone());
-                next
-            }
+    fn write_ts(&mut self) -> WriteTimestamp<T> {
+        let mut next = (self.next)();
+        if next.less_equal(&self.state.write_ts) {
+            next = self.state.write_ts.step_forward();
         }
+        assert!(self.state.read_ts.less_than(&next));
+        assert!(self.state.write_ts.less_than(&next));
+        self.state.write_ts = next.clone();
+        assert!(self.state.read_ts.less_equal(&self.state.write_ts));
+        let advance_to = next.step_forward();
+        WriteTimestamp {
+            timestamp: next,
+            advance_to,
+        }
+    }
+
+    /// Peek the current write timestamp.
+    fn peek_write_ts(&self) -> T {
```

Comment:
I read this and thought that maybe there was a different way to go from the previous design to here. I'll have to see how `peek_write_ts` gets used, but if it was just where we previously used `write_ts`, maybe there should be `write_ts` and `seal_write_ts`, or so.

Reply:
`peek_write_ts` is used where previously `peek_ts` was used. Group commit uses it to determine how long to sleep for in case the write timestamp is ahead of the system clock.